core/tee/tadb.c

```diff
@@ -738,7 +738,7 @@ TEE_Result tee_tadb_ta_read(struct tee_tadb_ta_read *ta, void *buf, size_t *len)
 			return res;
 	} else {
 		size_t num_bytes = 0;
-		size_t b_size = MIN(SIZE_4K, l);
+		size_t b_size = MIN(256U, l);
```

Comment:
> Impact on performance should be minimal as not much data are skipped in the payload.

Well, yes and no. It seems we always skip about 64K, due to the start offset of the first PT_LOAD segment.
Indeed, the segments in a TA look like so:

```
$ arm-linux-gnueabihf-readelf --program-header optee_test/out/ta/crypt/cb3e5ba0-adf1-11e0-998b-0002a5d5c51b.stripped.elf 

Elf file type is DYN (Shared object file)
Entry point 0x20
There are 5 program headers, starting at offset 52

Program Headers:
  Type           Offset   VirtAddr   PhysAddr   FileSiz MemSiz  Flg Align
  EXIDX          0x026ae0 0x00016ae0 0x00016ae0 0x00bd0 0x00bd0 R   0x4
  LOAD           0x010000 0x00000000 0x00000000 0x11c28 0x11c28 R E 0x10000
  LOAD           0x021c28 0x00011c28 0x00011c28 0x05df4 0x05df4 R   0x10000
  LOAD           0x028000 0x00018000 0x00018000 0x001dc 0x0ae58 RW  0x10000
  DYNAMIC        0x0277ec 0x000177ec 0x000177ec 0x00080 0x00080 RW  0x4

 Section to Segment mapping:
  Segment Sections...
   00     .ARM.exidx 
   01     .ta_head .text 
   02     .rodata .ARM.extab .ARM.extab.text.utee_panic .ARM.extab.text.__aeabi_ldivmod .ARM.extab.text.__aeabi_uldivmod .ARM.exidx .got .dynsym .rel.got .dynamic .dynstr .hash .rel.dyn 
   03     .data .bss 
   04     .dynamic 
```

Note the first PT_LOAD starts at file offset 0x10000.

Whereas, in a shared library (no linker script used), the first PT_LOAD segment starts at file offset 0x0. I suppose it would be the same in an executable if it weren't for the `.ta_head` section we're inserting first.

```
$ arm-linux-gnueabihf-readelf --program-header optee_examples/talibs/talib3/libtalib3.stripped.so 
Elf file type is DYN (Shared object file)
Entry point 0x288
There are 5 program headers, starting at offset 52

Program Headers:
  Type           Offset   VirtAddr   PhysAddr   FileSiz MemSiz  Flg Align
  EXIDX          0x0002bc 0x000002bc 0x000002bc 0x00010 0x00010 R   0x4
  LOAD           0x000000 0x00000000 0x00000000 0x002cc 0x002cc R E 0x10000
  LOAD           0x0002cc 0x000102cc 0x000102cc 0x00090 0x00090 RW  0x10000
  DYNAMIC        0x0002cc 0x000102cc 0x000102cc 0x00080 0x00080 RW  0x4
  GNU_STACK      0x000000 0x00000000 0x00000000 0x00000 0x00000 RW  0x10

 Section to Segment mapping:
  Segment Sections...
   00     .ARM.exidx 
   01     .hash .dynsym .dynstr .rel.plt .plt .text .rodata .ARM.exidx 
   02     .dynamic .got 
   03     .dynamic 
   04     
```

This is certainly not a serious issue, but there may be room for improvement. I guess we would be fine with a first loadable segment starting at offset 0, including the ELF header + the program headers then our `.ta_head` followed by `.text`?

Reply:
True :wink: